### PR TITLE
Sasha/exp2

### LINKED ISSUE #933: Stress testing the ssh output failed.
`ls / -laR | lolcat`

overwhelmed the server. I got a lot of these in the logs and also the web ui froze.

WARN[0537] Audit log cannot keep up. Dropping event ''   file=state/log.go:101 func=state.(*CachingAuditLog).post
WARN[0537] Audit log cannot keep up. Dropping event ''   file=state/log.go:101 func=state.(*CachingAuditLog).post
WARN[0537] Audit log cannot keep up. Dropping event ''   file=state/log.go:101 func=state.(*CachingAuditLog).post

### PR DESCRIPTION
**Pull request by klizhentas**
Fixes #933 

## Description

This PR contains implementation for the new log forwarder

Some important properties of this implementation:

* Introduces grpc and protobuf based serialization. My plan is to release grpc-based API in future versions 
  of teleport, however this time I only use protobuf as a payload encoding. I settled on gogoproto 
  marshaler as it's much faster (see benchmarks [here](https://github.com/alecthomas/go_serialization_benchmarks) )

* Without back-pressure on posting session chunks, audit log was loosing events
   because produce was much faster than consume and buffer was overflowing even if you set it to very
   large value, e.g. 10K . Back pressure is triggered on large outputs, e.g. `cat /dev/random` and usually corrects the output with `1ms` delays when log server flushes the output to disk, causing occasional 10-15ms latency spikes.

* Throttle is important to continue the session in case if audit log
  get slow, as the session output will block and timeout on every request

* It is important to pack chunks, because `ls / -laR | lolcat` would otherwise
   generate about 10K requests per second. With this packing approach
   we reduced this number to about 40-50 requests per second,
   we can now tweak this parameter now by setting queue size and flush buffers.

 * Current implementation attaches audit log forwarder per session. I have tested with 100 parallel sessions and the memory usage is about 100MB to make sure it's not too expensive

 * One can be concerned about channel and goroutine per audit forward, however I would not worry about that for now, as the overhead per channel is substantial, but not [critical](https://stackoverflow.com/questions/30767510/which-channel-type-uses-the-least-amount-of-memory-in-go)

* New implementation will make it easy to introduce "audit-mandatory" sessions that will freeze if the audit log can't be sent, however I did not bother to implement those.
 
### Thoughts

I spent several days doing benchmarks and various loadtests, as a result I have refined the dashboard and added various metrics outputs to teleport Prometheus endpoint. Here are some thoughts on the overall Teleport performance:

* We are creating lots of requests to auth server per interactive session that can hit local cache instead.
* We are creating lots of goroutines for interactive sessions (30 goroutines per session). 
* From the CPU standpoint, in case of many concurrent open/close sessions, handshakes are by far the most expensive (90% of the CPU time is taken by them)
* In case of `lolcat` - high CPU usage is mostly caused by SSH driven IO
  `gogoproto` completely removed the chunk encoding/decoding from the CPU profile picture, once I will migrate events stream to protobuf as well, this will be even less of an issue.
* My planned migration to auth server using GRPC will significantly improve our performance as SSH over HTTP is quite slow.

### CODE REVIEW

**Comment by klizhentas** [comment]
lolcat

**Comment by klizhentas** [comment]
lolcat

**Review by russjones — APPROVED** [review]

File: lib/events/auditlog.go [hunk 1]
```diff
@@ -249,22 +278,42 @@ func (l *AuditLog) migrateSessions() error {
...
-// PostSessionChunk writes a new chunk of session stream into the audit log
-func (l *AuditLog) PostSessionChunk(namespace string, sid session.ID, reader io.Reader) error {
-	if namespace == "" {
+// PostSessionChunks writes a new chunk of session stream into the audit log
+func (l *AuditLog) PostSessionSlice(slice SessionSlice) error {
+	if slice.Namespace == "" {
 		return trace.BadParameter("missing parameter Namespace")
 	}
-	sl, err := l.LoggerFor(namespace, sid)
+	if len(slice.Chunks) == 0 {
```
Comment (klizhentas) [inline lib/events/auditlog.go]: LoggerFor opens file, so did not want to open a file on server if we know it won't work

**Comment by klizhentas** [comment]
retest this please

**Comment by klizhentas** [comment]
retest this please

**Pull request merged**